mypyc/cstring.py

```diff
@@ -0,0 +1,53 @@
+"""Encode valid C string literals from Python strings.
+
+If a character is not allowed in C string literals, it is either emitted
+as an alphabetic escape sequence (e.g. '\\n'), or an octal escape
+sequence with exactly three digits ('\\oXXX'). Question marks are
+escaped to prevent trigraphs in the string literal from being
+interpreted (this escape sequence is invalid in Python).
+
+Consider the string literal 'AB\\xCDEF'. As one would expect, Python
+parses it as ['A', 'B', 0xCD, 'E', 'F']. However, the C standard
+specifies that all hexadecimal digits immediately following '\\x' will
+be interpreted as part of the escape sequence. Therefore, it is
+unexpectedly parsed as ['A', 'B', 0xCDEF].
+
+Emitting ('12\\x34' '56') would avoid this behaviour. However, we opt
+for simplicity and use octal escape sequences instead. They do not
+suffer from the same issue as they are defined to parse at most three
+octal digits.
+"""
+
+import string
+from typing import Tuple
+
+CHAR_MAP = ['\\{:03o}'.format(i) for i in range(256)]
+
+# It is safe to use string.printable as it always uses the C locale.
+for c in string.printable:
+    CHAR_MAP[ord(c)] = c
+
+# These assignments must come last because we prioritize the escape
+# sequences in the C standard over any other representation.
+CHAR_MAP[ord('\'')] = r'\''
```

Comment:
@msullivan This is the simplest method I have been able to devise to generate these escape sequences:

```python
for x in ("'", '"', '\\', 'a', 'b', 'f', 'n', 'r', 't', 'v'):
    escaped = '\\{}'.format(x)
    value = escaped.encode('ascii').decode('unicode_escape')
    C_CHAR_MAP[ord(value)] = escaped
```

While I am not a huge fan of manually listing each escape sequence, I dislike this method more.

Reply:
The encode/decode is a little hokey but I think I like this more

Reply:
The main reason I dislike this is that we're calling into Python's Unicode APIs. There's also `codecs.escape_decode` that handles ASCII only, but it's undocumented.